p4-16/spec/P4-16-spec.mdk

```diff
@@ -1503,6 +1503,47 @@ Each parameter may be labeled with a direction:
 	- an action parameter that can be set directly by another calling
       action; in this case it behaves like an `in` parameter
 
+Direction `out` parameters are always initialized at the beginning of
+execution of the portion of the program that has the `out` parameters,
+e.g. `control`, `parser`, `action`, function, etc.  This
```

Comment:
It is probably sufficient to avoid changing the meaning of existing P4_16 programs if this initialization were limited to only "top level controls/parsers" in an architecture, since the main example I know of where this initialization is expected is for the one direction `out` parameter that is typically a struct containing headers that is `out` from the parser, usually declared with a type parameter named `H`.

That restriction might be considered a bit odd in a language spec, perhaps, but I wanted to mention it for consideration.  "top level" in this situation could be defined as "an object given as a parameter to the package named `main`, or via any chain of package invocations starting from `main`".

Reply:
Not that I am opposed to it, but this proposed restriction pretty much admits that this little exception is a property of an architecture and not of the language, does it not? (And as a consequence this becomes a restriction on all the architectures, mandated by the language spec).

Reply:
If you have another idea of how to change the language spec in a way that _doesn't_ change the behavior of all those existing P4_16 programs that rely on `out` parameters being initialized to invalid, now is a good time to mention it :-)

Yes, this proposed way makes it a restriction on all `out` parameters everywhere.  We could restrict to only parsers, or only the top level parsers, etc. trying to limit the extra cost to an implementer, at the possible cost of people reading the spec and wondering what the heck it is restricted to that special case for.  The more restrictive we make it, the more I would propose having an appendix with a rationale for those future readers (and ourselves, after we forget).

Reply:
If you truly want to eliminate any and all requirements to initialize `out` parameters from the language spec completely (a slightly backwards-incompatible change at this point, mind you, but I don't mind going to P4_16 v2.x over it if no one else does), then ...

One way to do that is to change the architectures that have these `out` type `H` parameters for their parsers to `inout`.  Yes, I understand that breaks the compilation of all programs written for those architectures, until they are updated to change that one `out` in their parser definition to `inout`, to match the new architecture definition.

If you leave the architectures with direction `out`, there is no language-supported way for an architecture to initialize the value of that parameter from "outside of the parser" by providing an initial value, without basically saying: "I know this architecture says this parameter is `out`, but we fake it in our compiler and treat it like it is `inout` instead."

Reply:
@jafingerhut -- you are pretty much reading my thoughts. I agree that we should write this in an appendix or something. 

It sounds to me that the right thing to do if we would **not** worry about backwards compatibility is to declare that `out` is not initialized, periodm and use `inout` for most of the things that are passed into the top-level controls/parsers by (some typical) architectures. Generally, it matches the traditiona "header/metadata bus" concept well. 

However, we might need to bite the lip and save this for some bigger juncture, unless there is a big desire to do it right now (which I will not be against). The realist in me, though, says that stating that header-like `out` parameters will be pre-initialized is fine for now and especially fine if we limit those to the top-level parsers/controls only as you suggested.